contracts/v0.6/SimpleSwap.sol

```diff
@@ -0,0 +1,178 @@
+pragma solidity ^0.6.0;
+
+import "@chainlink/contracts/src/v0.6/Owned.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "./token/ERC677Receiver.sol";
+
+contract SimpleSwap is Owned, ReentrancyGuard {
+  using SafeMath for uint256;
+
+  event LiquidityUpdated(
+    uint256 indexed amount,
+    address indexed source,
+    address indexed target
+  );
+  event TokensSwapped(
+    uint256 indexed amount,
+    address indexed source,
+    address indexed target,
+    address caller
+  );
+
+  mapping(address => mapping(address => uint256)) private s_swappableAmount;
+
+  /**
+   * @dev deposits tokens from the target of a swap pair but does not return any
+   * @param amount count of liquidity being added
+   * @param source the token that can be swapped for what is being deposited
+   * @param target the token that can is being deposited for swapping
+   */
+  function addLiquidity(
+    uint256 amount,
+    address source,
+    address target
+  )
+    external
+  {
+    _addLiquidity(amount, source, target);
+
+    require(ERC20(target).transferFrom(msg.sender, address(this), amount), "transferFrom failed");
+  }
+
+  /**
+   * @dev withdraws tokens from the target of a swap pair.
+   * Only callable by owner
+   * @param amount count of liquidity being removed
+   * @param source the token that can be swapped for what is being removed
+   * @param target the token that can is being withdrawn from swapping
+   */
+  function removeLiquidity(
+    uint256 amount,
+    address source,
+    address target
+  )
+    external
+    onlyOwner()
+  {
+    _removeLiquidity(amount, source, target);
+
+    require(ERC20(target).transfer(msg.sender, amount), "transfer failed");
+  }
+
+  /**
+   * @dev exchanges the source token for target token
+   * @param amount count of tokens being swapped
+   * @param source the token that is being given
+   * @param target the token that is being taken
+   */
+  function swap(
+    uint256 amount,
+    address source,
+    address target
+  )
+    external
+    nonReentrant()
+  {
+    _removeLiquidity(amount, source, target);
+    _addLiquidity(amount, target, source);
```

Comment:
2 reasons for that. First, we can't reuse the methods in `onTokenTransfer` if we do that. Second, it breaks the checks/effects/interactions pattern in `swap`, which should be ok because we have the reentrancy guard.